src/hypothesis/extra/numpy.py

```diff
@@ -73,20 +96,206 @@ def is_scalar(spec):
+
+
+@st.defines_strategy
+def scalar_dtypes(integers=True, floats=True, times=True, strings=True):
+    strat = st.nothing()
+    if integers:
+        strat |= integer_dtypes()
+    if floats:
+        strat |= floating_dtypes()
+    if times:
+        strat |= timestamp_dtypes()
+    if strings:
+        strat |= timestamp_dtypes()
+    return strat
+
+
+@st.defines_strategy
+def integer_dtypes(signed=True, unsigned=True, bools=True,
+                   endianness='?', sizes=(8, 16, 32, 64)):
+    """Return a strategy for bools, unsigned int, and signed int dtypes, with
+    generation of subtypes controlled by the flag variables.
+
+    endianness may be ``<`` for little-endian, ``>`` for big-endian,
+    ``=`` for native byte order, or ``?`` to allow either byte order.
+    This argument only applies to dtypes of more than one byte.
+
+    sizes must be a collection of integer sizes in bits.  The default
+    (8, 16, 32, 64) covers the full range of sizes.
+
+    """
+    assert signed or unsigned or bools
+    assert endianness in ('?', '<', '=', '>')
+    if isinstance(sizes, int):
+        sizes = (sizes,)
+    assert sizes and all(s in (8, 16, 32, 64) for s in sizes)
+
+    def make_strat(dtype, flag):
+        if not flag:
+            return st.nothing()
+        if dtype == '?':
+            return st.just('?')
+        return st.builds(
+            dtype.format, st.sampled_from(endianness.replace('?', '<>')),
+            st.sampled_from(sorted(set(s // 8 for s in sizes))))
+
+    return (make_strat('?', bools) | make_strat('{}u{}', unsigned) |
+            make_strat('{}i{}', signed)).map(np.dtype)
+
+
+@st.defines_strategy
+def floating_dtypes(floats=True, complexs=True, real_sizes=(16, 32, 64)):
+    """Return a strategy for floating and complex dtypes.  Like
+    `integer_dtypes`, this strategy has a flag to disable each subtype.
+
+    real_sizes is the size in bits of the real part of a floating-point or
+    complex number.  A complex value will therefore be twice the size.
+
+    Complex halfs (a 16-bit real part) are not supported by numpy and will
+    not be generated by this strategy.
+    Larger floats (96 and 128 bit real parts) are not supported on all
+    platforms and therefore disabled by default.  To generate these dtypes,
+    include these values in the real_sizes argument.
+
+    """
+    assert real_sizes and all(s in (16, 32, 64, 96, 128) for s in real_sizes)
+    complex_sizes = [s * 2 for s in real_sizes if s != 16]
+    complexs &= bool(complex_sizes)
+
+    def make_strat(dtype, flag):
+        sizes = real_sizes if 'f' in dtype else complex_sizes
+        return st.sampled_from(sorted(set(s // 8 for s in sizes)))\
+            .map(dtype.format).map(np.dtype) if flag else st.nothing()
+
+    return make_strat('f{}', floats) | make_strat('c{}', complexs)
+
+
+@st.defines_strategy
+def timestamp_dtypes(max_period='Y', min_period='ns',
+                     allow_datetime=True, allow_timedelta=True):
```

Comment:
Fair enough.  I'll split these up, along with the numeric type strategies.  I think string types are confused often enough that it's useful to keep them in one strategy though.